usersession/userd/launcher.go

```diff
@@ -183,12 +189,190 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+	desktopFile, err := desktopFileIDToFilename(osutil.RegularFileExists, desktopFileID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	err = verifyDesktopFileLocation(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+  args = append([]string{"systemd-run", "--user", "--"}, args...)
+
+	cmd := exec.Command(args[0], args[1:]...)
+
+	if cmd.Run() != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+type fileExists func(string) bool
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+// Per https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id,
+// if desktop entries have dashes in the name ('-'), this could be an indication of subdirectories, so search
+// for those too. Eg, given foo-bar_baz_norf.desktop the following are searched for:
+//   o .../foo-bar_baz-norf.desktop
+//   o .../foo/bar_baz-norf.desktop
+//   o .../foo/bar_baz/norf.desktop
+//   o .../foo-bar_baz/norf.desktop
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(desktopFileExists fileExists, baseDir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(baseDir, strings.Join(splitFileId, "-"))
+
+	if desktopFileExists(desktopFile) {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (not all entries in $XDG_DATA_DIRS) and we know that snapd
+	// does not create subdirectories in that location.
+	for i := 1; i != len(splitFileId); i++ {
+		desktopFile, err := findDesktopFile(desktopFileExists, filepath.Join(baseDir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if err == nil {
+			return desktopFile, err
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileExists fileExists, desktopFileID string) (string, error) {
+
+	// OpenDesktopEntry() currently only supports launching snap applications from
+	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
+	// written by snapd and considered safe for userd to process.
+	// Since we only access /var/lib/snapd/desktop/applications, ignore
+	// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	baseDir := dirs.SnapDesktopFilesDir
```

Comment:
As mentioned earlier, this should be checking all applications directories found through `XDG_DATA_HOME` and `XDG_DATA_DIRS`.

Yes, we only want to accept desktop IDs provided by a file in `dirs.SnapDesktopFilesDir`, but we can't determine that unless we've checked everything that comes before it in the search path.